flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/ColumnsOperationExpander.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.expressions;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.typeinfo.IntegerTypeInfo;
+import org.apache.flink.table.api.TableException;
+import org.apache.flink.util.Preconditions;
+
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.table.expressions.BuiltInFunctionDefinitions.COLUMNS;
+import static org.apache.flink.table.expressions.BuiltInFunctionDefinitions.COLUMNS_INVERSE_SELECTION;
+import static org.apache.flink.table.expressions.BuiltInFunctionDefinitions.COLUMNS_RANGE;
+
+/**
+ * Expands column expressions to it's real parent's input references.
+ */
+@Internal
+public class ColumnsOperationExpander extends ApiExpressionDefaultVisitor<List<Expression>> {
```

Comment:
How about move `ColumnsOperationExpander` into `ApiExpressionUtils`?

Reply:
This should be another `ResolverRule` once #8062 is merged.

Reply:
@sunjincheng121 @dawidwys Thanks a lot for your advice. Maybe we can keep `ColumnsOperationExpander` here for now and replace it to a ResolverRule once #8062 is merged?

Reply:
Make sense for my side. That depends on which PR be merged first. another should rebase code.

Reply:
I would be strongly in favor merging this after #8062. This PR introduces a new complex feature. If we merge it before #8062 it immediately introduces technical debt.

Reply:
I think this whole class should be internal to the rule.

Reply:
This class is same with `LookupCallResolver` which is used not only in Rule but also used in TableImpl to make sure `extractFieldReferences` get the right results.